### PR TITLE
Generate CRDs by `controller-gen`

### LINKED ISSUE #2753: apiextensions.k8s.io/v1beta1/CustomResourceDefinition is deprecated
## Bug Report

**What version of Kubernetes are you using?**
`v1.16`

**What version of TiDB Operator are you using?**

`v1.1.0`

**What did you see instead?**
apiextensions.k8s.io/v1beta1/CustomResourceDefinition is deprecated by apiextensions.k8s.io/v1/CustomResourceDefinition and not supported by Kubernetes v1.19+ clusters.

### LINKED ISSUE #4032: apiextensions.k8s.io/v1beta1 CustomResourceDefinition is deprecated in v1.16+, unavailable in v1.22+; use apiextensions.k8s.io/v1 CustomResourceDefinition
## Bug Report
apiextensions.k8s.io/v1beta1 is deprecated, is there any plan to replace with v1?

**What version of Kubernetes are you using?**
1.20

**What did you do?**
kubectl apply -f https://raw.githubusercontent.com/pingcap/tidb-operator/v1.1.12/manifests/crd.yaml

**What did you expect to see?**
no warning messages

**What did you see instead?**
image

### LINKED ISSUE #4172: Operator projects using the removed APIs in k8s 1.22 requires changes.
### Problem Description

Kubernetes has been deprecating API(s), which will be removed and are no longer available in 1.22. Operators projects using [these APIs](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22) versions will **not** work on Kubernetes 1.22 or any cluster vendor using this Kubernetes version(1.22), such as OpenShift 4.9+. Following the APIs that are most likely your projects to be affected by:
- apiextensions.k8s.io/v1beta1: (Used for CRDs and available since v1.16)
- rbac.authorization.k8s.io/v1beta1: (Used for RBAC/rules and available since v1.8)
- admissionregistration.k8s.io/v1beta1 (Used for Webhooks and available since v1.16)

Therefore, looks like this project distributes solutions in the [repository](https://github.com/redhat-openshift-ecosystem/community-operators-prod/tree/main/operators/tidb-operator) and does not contain any version compatible with k8s 1.22/OCP 4.9. ([More info](https://github.com/redhat-openshift-ecosystem/community-operators-prod/discussions/138)). Following some findings by checking the distributions published:

- **tidb-operator.v1.1.6**: this distribution is using APIs which were deprecated and removed in v1.22. More info: https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22. Migrate the API(s) for CRD: (["backups.pingcap.com" "backupschedules.pingcap.com" "restores.pingcap.com" "tidbclusterautoscalers.pingcap.com" "tidbclusters.pingcap.com" "tidbinitializers.pingcap.com" "tidbmonitors.pingcap.com"]) 
- **tidb-operator.v1.1.7**: this distribution is using APIs which were deprecated and removed in v1.22. More info: https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22. Migrate the API(s) for CRD: (["backups.pingcap.com" "backupschedules.pingcap.com" "restores.pingcap.com" "tidbclusterautoscalers.pingcap.com" "tidbclusters.pingcap.com" "tidbinitializers.pingcap.com" "tidbmonitors.pingcap.com"]) 
- **tidb-operator.v1.1.2**: this distribution is using APIs which were deprecated and removed in v1.22. More info: https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22. Migrate the API(s) for CRD: (["backups.pingcap.com" "backupschedules.pingcap.com" "restores.pingcap.com" "tidbclusterautoscalers.pingcap.com" "tidbclusters.pingcap.com" "tidbinitializers.pingcap.com" "tidbmonitors.pingcap.com"]) 
- **tidb-operator.v1.1.5**: this distribution is using APIs which were deprecated and removed in v1.22. More info: https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22. Migrate the API(s) for CRD: (["backups.pingcap.com" "backupschedules.pingcap.com" "restores.pingcap.com" "tidbclusterautoscalers.pingcap.com" "tidbclusters.pingcap.com" "tidbinitializers.pingcap.com" "tidbmonitors.pingcap.com"]) 

**NOTE:** The above findings are only about the manifests shipped inside of the distribution. It is not checking the codebase.

### How to solve

It would be very nice to see new distributions of this project that are no longer using [these APIs](https://kubernetes.io/docs/reference/using-api/deprecation-guide/#v1-22) and so they can work on Kubernetes 1.22 and newer and published in the [community-operators](https://github.com/redhat-openshift-ecosystem/community-operators-prod/tree/main/operators/tidb-operator) collection. OpenShift 4.9, for example, will not ship operators anymore that do still use v1beta1 extension APIs.

Due to the number of options available to build Operators, it is hard to provide direct guidance on updating your operator to support Kubernetes 1.22. Recent versions of the [OperatorSDK](https://github.com/operator-framework/operator-sdk) greater than [1.0.0](https://github.com/operator-framework/operator-sdk/releases) and [Kubebuilder](https://github.com/kubernetes-sigs/kubebuilder) greater than [3.0.0](https://github.com/kubernetes-sigs/kubebuilder/releases) scaffold your project with the latest versions of these APIs (all that is generated by tools only). See the guides to upgrade your projects with OperatorSDK [Golang](https://sdk.operatorframework.io/docs/building-operators/golang/migration/), [Ansible](https://sdk.operatorframework.io/docs/building-operators/ansible/migration/), [Helm](https://sdk.operatorframework.io/docs/building-operators/helm/migration/) or the [Kubebuilder](https://book.kubebuilder.io/migration/v2vsv3.html) one. For APIs other than the ones mentioned above, you will have to check your code for usage of removed API versions and upgrade to newer APIs. The details of this depend on your codebase.

**If this projects only need to migrate the API for CRDs and it was built with [OperatorSDK](https://github.com/operator-framework/operator-sdk) versions lower than [1.0.0](https://github.com/operator-framework/operator-sdk/releases) then, you maybe able to solve it with an [OperatorSDK](https://github.com/operator-framework/operator-sdk) version >= v0.18.x < 1.0.0:**

> $ operator-sdk generate crds --crd-version=v1
> INFO[0000] Running CRD generator.                       
> INFO[0000] CRD generation complete.

**Alternatively, you can try to upgrade your manifests with [controller-gen](https://book.kubebuilder.io/reference/controller-gen.html) ([version >= v0.4.1](https://github.com/kubernetes-sigs/controller-tools/releases/tag/v0.4.1)) :**

#### If this project does not use Webhooks:

> $ controller-gen crd:trivialVersions=true,preserveUnknownFields=false rbac:roleName=manager-role  paths="./..."

#### If this project is using Webhooks:

1. Add the markers [sideEffects](https://github.com/kubernetes-sigs/controller-tools/blob/master/pkg/webhook/parser.go#L81) and [admissionReviewVersions](https://github.com/kubernetes-sigs/controller-tools/blob/master/pkg/webhook/parser.go#L114) to your webhook (Example with sideEffects=None and admissionReviewVersions={v1,v1beta1}: [memcached-operator/api/v1alpha1/memcached_webhook.go](https://github.com/operator-framework/operator-sdk/blob/master/testdata/go/v3/memcached-operator/api/v1alpha1/memcached_webhook.go#L39)):

2. Run the command:
> $ controller-gen crd:trivialVersions=true,preserveUnknownFields=false rbac:roleName=manager-role webhook paths="./..."

For further information and tips see the [comment](https://github.com/redhat-openshift-ecosystem/community-operators-prod/discussions/138#discussioncomment-1245759).

### PR DESCRIPTION
**Pull request by KanShiori**
### What problem does this PR solve?

Fix #2753 Fix #4032 Fix #4172

Ref #3878

### What is changed and how does it work?

* use `controller-gen` to generate CRDs, and modify the related script
* modify some code to make `controller-gen` happy
* generate v1 CRDs and v1beta CRDs in dir `manifest/crd`, and merge them into the files `manifest/crd.yaml` and `manifest/crd_v1beta1.yaml`

### Code changes

- [x] Has Go code change
- [x] Has CI related scripts change

### Tests

- [ ] Unit test 
- [ ] E2E test 
- [x] Manual test 

GKE 1.20.8
* **`manifest/crd_v1beta1.yaml` is incompatible**, ref: https://github.com/kubernetes/kubernetes/issues/91395
* Apply `manifest/crd.yaml` and deploy
   - [x] TiDBCluster
   - [x] TiDBInitailizer
   - [x] TiDBMonitor
   - [x] DMCluster
   - [x] Backup Restore

GKE 1.17.17 / EKS 1.16.15
* Apply `manifest/crd_v1beta1.yaml` and deploy 
   - [x] TiDBCluster
   - [x] TiDBInitailizer
   - [x] TiDBMonitor
   - [x] DMCluster
   - [x] Backup Restore
* Apply `manifest/crd.yaml` and deploy
   - [x] TiDBCluster
   - [x] TiDBInitailizer
   - [x] TiDBMonitor
   - [x] DMCluster
   - [x] Backup Restore

CRD upgrade
- [x] Deploy v1beta1 CRD and upgrade to v1 CRD by `kubectl replace`

- [ ] No code 

### Side effects

- [ ] Breaking backward compatibility 
- [ ] Other side effects: 

### Related changes

- [ ] Need to cherry-pick to the release branch 
- [ ] Need to update the documentation 

### Release Notes

Please refer to [Release Notes Language Style Guide](https://github.com/pingcap/tidb-operator/blob/master/docs/release-note-guide.md) before writing the release note.

```release-note

```

### CODE REVIEW

File: pkg/apis/pingcap/v1alpha1/types.go [hunk 1]
```diff
@@ -351,7 +370,7 @@ type PDSpec struct {
 	ServiceAccount string `json:"serviceAccount,omitempty"`
 
 	// The desired ready replicas
-	// +kubebuilder:validation:Minimum=1
+	// +kubebuilder:validation:Minimum=0
```
Comment (csuzhangxc) [inline pkg/apis/pingcap/v1alpha1/types.go:373]: can it be `0` (for a heterogeneous cluster?)

**Review by csuzhangxc — COMMENTED** [review]
how many diffs between old `manifests/crd.yaml` and new `manifests/crd_v1beta1.yaml`?

**Comment by KanShiori** [comment]
> how many diffs between old `manifests/crd.yaml` and new `manifests/crd_v1beta1.yaml`?

`manifests/crd_v1beta1.yaml` add `required` field, if the json tag of property don't contain `omitempty`.

`manifests/crd_v1beta1.yaml` contains `status` field, that is removed in old `manifests/crd.yaml`.

I think the others are the same.

**Comment by csuzhangxc** [comment]
> > how many diffs between old `manifests/crd.yaml` and new `manifests/crd_v1beta1.yaml`?
> 
> `manifests/crd_v1beta1.yaml` add `required` field, if the json tag of property don't contain `omitempty`.
> 
> `manifests/crd_v1beta1.yaml` contains `status` field, that is removed in old `manifests/crd.yaml`.
> 
> I think the others are the same.

I diffed `tidbclusters` locally, it has many properties diffs, but that should be harmless.

**Review by csuzhangxc — APPROVED** [review]

**Review by liubog2008 — APPROVED** [review]
LGTM

**Review by DanielZhangQD — APPROVED** [review]
LGTM

**Pull request merged**